sql: functions have EXECUTE priv for public by default

rafiss opened the pull request:
This uses the default privilege infrastructure in a similar way to how the USAGE privilege for types is given to `public` by default.

This also fixes a display bug that was preventing the ALL privileges that are given to root and admin from being shown in SHOW.

informs https://github.com/cockroachdb/cockroach/issues/103842
A complete fix for that issue may also require an upgrade migration to give existing
functions the EXECUTE privilege for public.

Release note (sql change): The public pseudo-role now receives the EXECUTE privilege by default for all user-defined functions that are created. This can be adjusted by using ALTER DEFAULT PRIVILEGES.

rafiss committed:
sql: functions have EXECUTE priv for public by default

This uses the default privilege infrastructure in a similar way to how
the USAGE privilege for types is given to `public` by default.

This also fixes a display bug that was preventing the ALL privileges
that are given to root and admin from being shown in SHOW.

Release note (sql change): The public pseudo-role now receives the
EXECUTE privilege by default for all user-defined functions that are
created. This can be adjusted by using ALTER DEFAULT PRIVILEGES.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107317)
<!-- Reviewable:end -->

chengxiong-ruan reviewed (approved):
looks straightfoward to me. nice work!

rafiss committed:
sql: small fixes to nameconcatoid

There was an edge case if the result was exactly 64 characters where the
input would not be truncated correctly. This also makes the
information_schema.role_routine_grants table use the same logic for
computing the specific_name.

Release note: None

rafiss commented:
tftr!

bors r+

mgartner commented:
Thanks for fixing this! Note that I'm still working on #103842 to fix EXECUTE privileges in general.
